db/crud.go

```diff
@@ -1476,63 +1470,50 @@ func (db *DatabaseCollectionWithUser) storeOldBodyInRevTreeAndUpdateCurrent(ctx
 	metaMap, err = doc.GetMetaMap(db.userXattrKey())
 	if err != nil {
 		return
 	}
 
-	mutableBody, err = newDoc.GetDeepMutableBody()
+	revBody, err = newDoc.BodyBytes()
 	if err != nil {
 		return
 	}
 
-	err = validateNewBody(mutableBody)
+	err = validateNewBody(newDoc.Body())
 	if err != nil {
 		return
 	}
 
 	newRevID = newDoc.RevID
-
-	mutableBody[BodyId] = doc.ID
-	mutableBody[BodyRev] = newRevID
-	if newDoc.Deleted {
-		mutableBody[BodyDeleted] = true
-	}
-
 	return
 }
 
 // Run the sync function on the given document and body. Need to inject the document ID and rev ID temporarily to run
 // the sync function.
-func (db *DatabaseCollectionWithUser) runSyncFn(ctx context.Context, doc *Document, body Body, metaMap map[string]interface{}, newRevId string) (*uint32, string, base.Set, channels.AccessMap, channels.AccessMap, error) {
-	channelSet, access, roles, syncExpiry, oldBody, err := db.getChannelsAndAccess(ctx, doc, body, metaMap, newRevId)
+func (db *DatabaseCollectionWithUser) runSyncFn(ctx context.Context, doc *Document, body []byte, metaMap channels.MetaMap, newRevId string, deleted bool) (*uint32, string, base.Set, channels.AccessMap, channels.AccessMap, error) {
+	channelSet, access, roles, syncExpiry, oldBody, err := db.getChannelsAndAccess(ctx, doc, body, metaMap, newRevId, deleted)
 	if err != nil {
 		return nil, ``, nil, nil, nil, err
 	}
 	db.checkDocChannelsAndGrantsLimits(ctx, doc.ID, channelSet, access, roles)
 	return syncExpiry, oldBody, channelSet, access, roles, nil
 }
 
-func (db *DatabaseCollectionWithUser) recalculateSyncFnForActiveRev(ctx context.Context, doc *Document, metaMap map[string]interface{}, newRevID string) (channelSet base.Set, access, roles channels.AccessMap, syncExpiry *uint32, oldBodyJSON string, err error) {
+func (db *DatabaseCollectionWithUser) recalculateSyncFnForActiveRev(ctx context.Context, doc *Document, metaMap channels.MetaMap, newRevID string) (channelSet base.Set, access, roles channels.AccessMap, syncExpiry *uint32, oldBodyJSON string, err error) {
 	// In some cases an older revision might become the current one. If so, get its
 	// channels & access, for purposes of updating the doc:
 	curBodyBytes, err := db.getAvailable1xRev(ctx, doc, doc.CurrentRev)
 	if err != nil {
 		return
 	}
 
-	var curBody Body
-	err = curBody.Unmarshal(curBodyBytes)
-	if err != nil {
-		return
-	}
-
-	if curBody != nil {
+	if curBodyBytes != nil && string(curBodyBytes) != "null" {
 		base.DebugfCtx(ctx, base.KeyCRUD, "updateDoc(%q): Rev %q causes %q to become current again",
 			base.UD(doc.ID), newRevID, doc.CurrentRev)
-		channelSet, access, roles, syncExpiry, oldBodyJSON, err = db.getChannelsAndAccess(ctx, doc, curBody, metaMap, doc.CurrentRev)
+		channelSet, access, roles, syncExpiry, oldBodyJSON, err = db.getChannelsAndAccess(ctx, doc, curBodyBytes, metaMap, doc.CurrentRev, false /*TODO: Is false correct??*/)
```

Comment:
address this TODO?

Reply:
Ah, right. `false` is correct, so I'll remove the comment.